core/engine.go

```diff
@@ -436,25 +473,25 @@ func (e *Engine) processSamplesForMetrics(sampleContainers []stats.SampleContain
 		}
 
 		for _, sample := range samples {
-			m, ok := e.Metrics[sample.Metric.Name]
-			if !ok {
-				m = stats.New(sample.Metric.Name, sample.Metric.Type, sample.Metric.Contains)
-				m.Thresholds = e.thresholds[m.Name]
-				m.Submetrics = e.submetrics[m.Name]
+			m := sample.Metric // this should have come from the Registry, no need to look it up
+			if !m.Observed {
+				// But we need to add it here, so we can show data in the
+				// end-of-test summary for this metric
 				e.Metrics[m.Name] = m
+				m.Observed = true
```

Comment:
I really dislike this new property on the Metric type. I would really prefer if we go back to the map. 

I do think Metric already has too many responsibilities and now adding it knowing it was "observed" seems contra productive. 
we even still add it to the map so ... :shrug: Why not just get it out of it.

Reply:
I really dislike it as well, I want to completely split apart the current `Metric` into at least two parts:
- The first part, which should be the actual `Metric`, will be just the `Name`, `Type` and Contains` - that should be the only thing that the `Registry` actually controls and cares about
- A second completely separate `struct` should be the `Sink` and `Observed` and probably the threshold and submetric things. That will only be a concern of the `MetricsRegistry`. I don't think they should even be exported... In any case, the mapping from a `*Metric` to that second struct will happen only in the `MetricsRegistry`.

Unfortunately, until we have that separation, I kind of had to put `Observed` in the `Metric` struct, since that's where the `Sink` already was :disappointed: Doing otherwise without even more refactoring would cause us to have a map lookup for every sample of every metric that k6 generates... :disappointed: I intended on splitting this when I work on the time series refactoring.

Reply:
btw this is what I mean by this TODO: https://github.com/grafana/k6/blob/f06d6f0b673ebc9884995478a4f8d5d7ac03a923/stats/stats.go#L449-L450

Reply:
> e a map lookup for every sample of every metric that k6 generates..

Didn't we have the same up to this point as well?

Reply:
Yes, we did, but we don't actually need to have that, so I guess you can consider it an optimization? :sweat_smile: : 

I probably dislike the extra item in `Metric` even more than you, but with `Sink` already in there, this actually seems a better (temporary and transient) architecture to me, since all of this state is now kept in `Metric` and will be easy to refactor into a different struct all at the same time. Previously, the `Observed` state was implicitly tracked by that map in the `Engine`, while all of the other state was in every `Metric`...

If you want, I am fine with implementing my `TODO` above now, before we merge this PR, though I'd prefer to do it in a separate PR after this https://github.com/grafana/k6/pull/2442 PR from @oleiade is also merged. It will leave the `metrics.Registry` as the only source of truth for metrics, so we'll be easily able to assign them sequential IDs and use that in the `MetricsEngine` without the need for a map.